external-plugins/referee/server/server.go

```diff
@@ -0,0 +1,187 @@
+}
+
+// Server implements http.Handler. It validates incoming GitHub webhooks and
+// then dispatches them to the appropriate plugins.
+type Server struct {
+	TokenGenerator func() []byte
+	BotName        string
+
+	Log *logrus.Entry
+
+	GithubClient    githubClient
+	GHGraphQLClient ghgraphql.GitHubGraphQLClient
+
+	// DryRun says whether to create comments on PRs or to just write them to the log
+	DryRun bool
+
+	// Team is the name of the GitHub team that should be pinged
+	Team string
+}
+
+// ServeHTTP validates an incoming webhook and puts it into the event channel.
+func (s *Server) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	eventType, eventGUID, payload, ok, _ := github.ValidateWebhook(w, r, s.TokenGenerator)
+	if !ok {
+		return
+	}
+
+	if err := s.handleEvent(eventType, eventGUID, payload); err != nil {
+		s.Log.WithError(err).Error("Error parsing event.")
+	}
+}
+
+func (s *Server) handleEvent(eventType, eventGUID string, payload []byte) error {
+	l := logrus.WithFields(
+		logrus.Fields{
+			"event-type":     eventType,
+			github.EventGUID: eventGUID,
+		},
+	)
+	switch eventType {
+	//https://docs.github.com/de/webhooks/webhook-events-and-payloads#issue_comment
+	case "issue_comment":
+		var ic github.IssueCommentEvent
+		if err := json.Unmarshal(payload, &ic); err != nil {
+			return err
+		}
+		go func() {
+			if err := s.handlePullRequestComment(ic); err != nil {
+				s.Log.WithError(err).WithFields(l.Data).Errorf("%s failed.", PluginName)
+			}
+		}()
+	default:
+		s.Log.WithFields(l.Data).Debugf("skipping event of type %q", eventType)
+	}
+	return nil
+}
+
+func (s *Server) handlePullRequestComment(ic github.IssueCommentEvent) error {
+	org := ic.Repo.Owner.Login
+	repo := ic.Repo.Name
+	num := ic.Issue.Number
+	action := ic.Action
+	user := ic.Comment.User.Login
+
+	switch action {
+	case github.IssueCommentActionCreated:
+	default:
+		s.Log.Debugf("skipping referee for action %s on pull request %s/%s#%d by %s", action, org, repo, num, user)
+		return nil
+	}
+
+	if !ic.Issue.IsPullRequest() {
+		s.Log.Debugf("skipping referee since %s/%s#%d is not a pull request", org, repo, num)
+		return nil
+	}
+
+	if !pjutil.RetestRe.MatchString(ic.Comment.Body) &&
+		!pjutil.RetestRequiredRe.MatchString(ic.Comment.Body) &&
+		!pjutil.TestAllRe.MatchString(ic.Comment.Body) {
+		s.Log.Debugf("skipping referee since %s/%s#%d comment didn't contain command triggering tests", org, repo, num)
+		return nil
+	}
+
+	prTimeLineForLastCommit, err := s.GHGraphQLClient.FetchPRTimeLineForLastCommit(org, repo, num)
+	if err != nil {
+		s.Log.Fatalf("failed to fetch number of retest comments for pr %s/%s#%d: %v", org, repo, num, err)
+	}
+
+	if prTimeLineForLastCommit.NumberOfRetestComments < 5 {
+		s.Log.Debugf("skipping referee due to less number of retest comments for pr %s/%s#%d: %v", org, repo, num, prTimeLineForLastCommit)
+		return nil
+	}
+
+	if prTimeLineForLastCommit.WasHeld {
```

Comment:
Actually I wasn't sure about this? Initially I thought we should look at whether the hold was cancelled, but since IIRC we discussed that we only care if a hold had been placed at some point, and then disregard the PR anyway, I thought we'd be ok here - especially i.e. like in the case where a hold has initially been placed and then removed by a contributor, we don't want to hold again.

Reply:
if hold was added manually just for other reasons, and then was removed manually, aint the plugin should now do hold in case the strike was big enough?

Reply:
Maybe we need to look specifically for this bot hold - if it has occurred once then we don't look at that PR again for other retests.

Reply:
possible to add a new hold label, which will be used only by the bot if it helps a bit